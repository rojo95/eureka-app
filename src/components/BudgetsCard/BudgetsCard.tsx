import {
    Image,
    StyleSheet,
    TouchableOpacity,
    TouchableOpacityProps,
    View,
} from "react-native";
import Text from "../Text/Text";
import { Badge, DefaultTheme, useTheme } from "react-native-paper";

interface BudgetsCardInterface extends TouchableOpacityProps {
    index?: any;
    description: any;
    status: { id: number; name: string };
    costo: any;
    venta: any;
    onPress?: () => void;
}

const BudgetsCard = ({
    index,
    description,
    status,
    costo,
    venta,
    onPress,
}: BudgetsCardInterface) => {
    const theme: DefaultTheme = useTheme();
    const styles = StyleSheet.create({
        container: {
            flexDirection: "row",
            borderTopLeftRadius: 0,
            borderBottomStartRadius: 0,
            borderRadius: 10,
            overflow: "hidden",
            backgroundColor: theme.colors.primaryContrast,
            shadowColor: "#000",
            elevation: 5,
            marginVertical: 10,
            marginEnd: 10,
            alignSelf: "center",
        },
        content: {
            flex: 1,
            padding: 16,
        },
        description: {
            fontSize: 18,
            fontWeight: "bold",
            marginBottom: 8,
            color: theme.colors.dark,
        },
        number: {
            fontSize: 14,
            textAlign: "right",
        },
        code: {
            fontSize: 14,
            color: theme.colors.codeColor,
        },
        badgeContainer: {
            flexDirection: "row",
            alignItems: "center",
            marginBottom: 8,
        },
        badge: {
            marginRight: 8,
            color: theme.colors.dark,
        },
        costo: {
            color: theme.colors.danger,
            fontWeight: "bold",
        },
        venta: {
            color: theme.colors.success,
            fontWeight: "bold",
        },
    });

    const CustomBadge: React.FC<{ id: number; name: string }> = ({
        id,
        name,
    }) => {
        const info = {
            ...(id === 1
                ? { color: theme.colors.primaryLight }
                : id === 2
                ? { color: theme.colors.deepBlueLight }
                : id === 4
                ? { color: theme.colors.dangerLight }
                : { color: theme.colors.successLight }),
        };
        return (
            <View style={styles.badgeContainer}>
                <Badge style={[styles.badge, { backgroundColor: info.color }]}>
                    {name}
                </Badge>
            </View>
        );
    };

    return (
        <TouchableOpacity
            style={{ flex: 1, alignItems: "center", justifyContent: "center" }}
            onPress={onPress}
        >
            <View>
                <View style={styles.container}>
                    <View style={styles.content}>
                        <View>
                            <Text style={styles.code}>{index}</Text>
                        </View>
                        <Text style={styles.description}>{description}</Text>
                        <CustomBadge id={status?.id} name={status?.name} />
                        <Text style={styles.number}>
                            <Text style={styles.costo}>{costo}€</Text> -{" "}
                            <Text style={styles.venta}>{venta}€</Text>
                        </Text>
                    </View>
                </View>
            </View>
        </TouchableOpacity>
    );
};

export default BudgetsCard;
