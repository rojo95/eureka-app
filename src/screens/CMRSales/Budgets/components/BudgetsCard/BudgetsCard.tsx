import {
    StyleSheet,
    TouchableOpacityProps,
    View,
    TouchableOpacity,
} from "react-native";
import Text from "../../../../../components/Text/Text";
import { ActivityIndicator, DefaultTheme, useTheme } from "react-native-paper";
import CustomBadge from "../../../../../components/CustomBadge/CustomBadge";
import { formatPrices } from "../../../../../utils/numbers";
import { useContext, useEffect, useState } from "react";
import { getColorState } from "../../utils/utils";
import { UserContext } from "../../../../../contexts/UserContext";

interface BudgetsCardInterface extends TouchableOpacityProps {
    number?: string;
    title: string;
    state: { id: number; name: string };
    totalCost: number;
    totalSale: number;
    onPress?: () => void;
}

const BudgetsCard = ({
    number,
    title,
    state,
    totalCost,
    totalSale,
    onPress,
}: BudgetsCardInterface) => {
    const { language } = useContext(UserContext);
    const theme: DefaultTheme = useTheme();
    const [formattedCost, setFormattedCost] = useState<string>("");
    const [formattedSale, setFormattedSale] = useState<string>("");

    const themedStyles = StyleSheet.create({
        container: {
            backgroundColor: theme.colors.primaryContrast,
        },
        title: {
            color: theme.colors.dark,
        },
        code: {
            color: theme.colors.codeColor,
        },
        totalCost: {
            color: theme.colors.danger,
        },
        totalSale: {
            color: theme.colors.success,
        },
    });

    useEffect(() => {
        (() => {
            const formattedCost = formatPrices({
                number: totalCost || 0,
                language,
            });
            const formattedSale = formatPrices({
                number: totalSale || 0,
                language,
            });

            setFormattedCost(formattedCost);
            setFormattedSale(formattedSale);
        })();
    }, [totalCost, totalSale]);

    return (
        <TouchableOpacity
            style={{ flex: 1, alignItems: "center", justifyContent: "center" }}
            onPress={onPress}
        >
            <View style={[styles.container, themedStyles.container]}>
                <View style={[styles.content]}>
                    <View>
                        <Text style={[styles.code, themedStyles.code]}>
                            {number}
                        </Text>
                    </View>
                    <Text style={[styles.bold, themedStyles.title]}>
                        {title}
                    </Text>
                    <CustomBadge
                        customStyles={{
                            backgroundColor: getColorState({
                                statusId: state.id,
                                theme,
                            }),
                            color: theme.colors.dark,
                        }}
                    >
                        {state.name}
                    </CustomBadge>
                    <Text style={[styles.number]}>
                        {formattedCost === "" ? (
                            <ActivityIndicator size="small" />
                        ) : (
                            <Text>
                                <Text
                                    style={[
                                        styles.bold,
                                        themedStyles.totalCost,
                                    ]}
                                >
                                    {formattedCost}€
                                </Text>{" "}
                                -{" "}
                                <Text
                                    style={[
                                        styles.bold,
                                        themedStyles.totalSale,
                                    ]}
                                >
                                    {formattedSale}€
                                </Text>
                            </Text>
                        )}
                    </Text>
                </View>
            </View>
        </TouchableOpacity>
    );
};

export default BudgetsCard;

const styles = StyleSheet.create({
    container: {
        flexDirection: "row",
        borderTopLeftRadius: 0,
        borderBottomStartRadius: 0,
        borderRadius: 10,
        overflow: "hidden",
        shadowColor: "#000",
        elevation: 5,
        marginVertical: 10,
        marginEnd: 10,
        alignSelf: "center",
    },
    content: {
        flex: 1,
        padding: 16,
    },
    title: {
        fontSize: 18,
        marginBottom: 8,
    },
    number: {
        fontSize: 14,
        textAlign: "right",
    },
    code: {
        fontSize: 14,
    },
    bold: {
        fontWeight: "bold",
    },
});
