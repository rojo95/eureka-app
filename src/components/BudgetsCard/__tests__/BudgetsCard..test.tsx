import React from "react";
import renderer from "react-test-renderer";
import BudgetsCard from "../BudgetsCard";

describe("Test to evaluate the <BudgetsCard/> component", () => {
    it("should render correctly", () => {
        const tree: any = renderer.create(<BudgetsCard />).toJSON();
        expect(tree.children.length).toBe(1);
    });

    it("should rendeer correctly the cost value", () => {
        const testCosto = 500;

        const tree: any = renderer
            .create(<BudgetsCard costo={testCosto} />)
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[3]
                ?.children[0].children;
        expect(componentBase).toEqual([`${testCosto}`, "€"]);
    });

    it("should rendeer correctly the sell value", () => {
        const testVenta = 600;

        const tree: any = renderer
            .create(<BudgetsCard venta={testVenta} />)
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[3]
                ?.children[3].children;
        expect(componentBase).toEqual([`${testVenta}`, "€"]);
    });

    it("should rendeer correctly the status text", () => {
        const testStatusText = "En progreso";

        const tree: any = renderer
            .create(
                <BudgetsCard
                    status={{ id: 1, name: testStatusText }}
                />
            )
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[2]
                ?.children[0]?.children[0];
        expect(componentBase).toContain(testStatusText);
    });

    it("should rendeer correctly the index code", () => {
        const idTest = "1234";

        const tree: any = renderer
            .create(<BudgetsCard index={idTest} />)
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[0]
                ?.children[0]?.children[0];
        expect(componentBase).toContain(idTest);
    });

    it("should rendeer correctly the description", () => {
        const testDesc = "Descripcion de prueba";

        const tree: any = renderer
            .create(<BudgetsCard description={testDesc} />)
            .toJSON();

        const componentBase =
            tree?.children[0]?.children[0]?.children[0]?.children[1]?.children;
        expect(componentBase).toContain(testDesc);
    });

    it("render card without change", () => {
        const tree = renderer.create(<BudgetsCard />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
