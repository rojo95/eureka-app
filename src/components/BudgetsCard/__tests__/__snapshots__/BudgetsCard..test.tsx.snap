// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test to evaluate the <BudgetsCard/> component render card without change 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
>
  <View>
    <View
      style={
        {
          "alignSelf": "center",
          "backgroundColor": undefined,
          "borderBottomStartRadius": 0,
          "borderRadius": 10,
          "borderTopLeftRadius": 0,
          "elevation": 5,
          "flexDirection": "row",
          "marginEnd": 10,
          "marginVertical": 10,
          "overflow": "hidden",
          "shadowColor": "#000",
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "padding": 16,
          }
        }
      >
        <View>
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": undefined,
                  "fontSize": 14,
                },
              ]
            }
          />
        </View>
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "color": undefined,
                "fontSize": 18,
                "fontWeight": "bold",
                "marginBottom": 8,
              },
            ]
          }
        />
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "marginBottom": 8,
            }
          }
        >
          <Text
            collapsable={false}
            numberOfLines={1}
            style={
              {
                "alignSelf": "flex-end",
                "backgroundColor": "rgba(179, 38, 30, 1)",
                "borderRadius": 10,
                "color": undefined,
                "fontSize": 10,
                "height": 20,
                "lineHeight": 10,
                "marginRight": 8,
                "minWidth": 20,
                "opacity": 1,
                "overflow": "hidden",
                "paddingHorizontal": 3,
                "textAlign": "center",
                "textAlignVertical": "center",
              }
            }
          />
        </View>
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "fontSize": 14,
                "textAlign": "right",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": undefined,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            €
          </Text>
           -
           
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "color": undefined,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            €
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
